Unit 2.1 - Convnext final to main (#197)
johko/computer-vision-course: chapters/en/Unit 2 - Convolutional Neural Networks/googlenet.mdx

# GoogLeNet

> ### 1. Overview

Reviewer: ```suggestion
## Overview
```

> The Inception architecture, a type of convolutional neural network (CNN) designed for tasks in computer vision such as classification and detection, stands out due to its efficiency. Containing fewer than 7 million parameters, it is significantly more compact than its predecessors, being 9 times smaller than AlexNet and 22 times smaller than VGG16. This architecture gained recognition in the ImageNet 2014 challenge, where Google's adaptation, named GoogLeNet (a tribute to LeNet), set new benchmarks in performance while utilizing fewer parameters compared to previous leading methods.

Reviewer: ```suggestion
The Inception architecture, a convolutional neural network (CNN) designed for tasks in computer vision such as classification and detection, stands out due to its efficiency. It contains fewer than 7 million parameters and is significantly more compact than its predecessors, being 9 times smaller than AlexNet and 22 times smaller than VGG16. This architecture gained recognition in the ImageNet 2014 challenge, where Google's adaptation, named GoogLeNet (a tribute to LeNet), set new benchmarks in performance while utilizing fewer parameters compared to previous leading methods.
```

> ### 2. Architectural Innovations

Reviewer: ```suggestion
### Architectural Innovations
```

> ### 2. Architectural Innovations

Reviewer: ```suggestion
## Architectural Innovations
```

Before the advent of the Inception architecture, models like AlexNet and VGG demonstrated the benefits of deeper network structures. However, deeper networks typically entail more computational steps and can lead to issues such as overfitting and the vanishing gradient problem. The Inception architecture offers a solution, enabling the training of complex CNNs with a reduced count of floating-point parameters.

> In conventional CNN design, layers are typically categorized as either pooling or convolution layers, with specific sizes for convolution filters. Although layering different sizes of convolution filters is beneficial for various tasks, it can rapidly increase the total number of parameters. The Inception architecture takes a different approach by running convolution filters of various sizes (1x1, 3x3, 5x5) in parallel. These are then integrated with max pooling into a unified component known as the Inception module. The GoogLeNet architecture is composed of a series of 9 such Inception modules. This configuration allows the network to maintain flexibility and learn complex tasks without a substantial increase in depth.

Reviewer 2: ```suggestion
In a conventional CNN design, layers are typically categorized as either pooling or convolution layers, with specific sizes for convolution filters. Although layering different sizes of convolution filters is beneficial for various tasks, it can rapidly increase the total number of parameters. The Inception architecture takes a different approach by running the convolution filters of various sizes (1x1, 3x3, 5x5) in parallel.  That means it is possible to get different lower-dimensional embeddings -and hence, more information- from the same higher-dimensional features using these parallel processes! These are then integrated with max pooling into a unified component known as the Inception module. The GoogLeNet architecture is composed of a series of 9 such Inception modules. This configuration allows the network to maintain flexibility and learn complex tasks without a substantial increase in depth.
```

## Code 

> 

Reviewer: Would be nice to add an explanation here

Reviewer 2: @merveenoyan added explanation

> ### 3. Code 

Reviewer: ```suggestion
## Code 
```

> import torch

Reviewer: Would be nice to enclose it with three backticks python (```python)

Reviewer 3: @sezan92  thanks for driving this. Merve has few changes suggested, can you please fix it?

> import torch

Reviewer 2: ```suggestion
```python
import torch
```

> import torch.nn as nn

Reviewer 3: Do you want to move this code to notebook and put a link here?

Reviewer 2: @ratan is that new requirement? currently we are fine with mdx. if that is required we can do in later PRs

Reviewer 3: @sezan92 that was a suggestion. Otherwise it looks good.

Reviewer 2: @ratan thanks for clarifying. please approve in that case ^_^
